x/bsonx/bsoncore/bsonbuilder.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+// Package bsoncore contains functions that can be used to encode and decode bson
+// elements and values to or from a slice of bytes. These functions are aimed at
+// allowing low level manipulation of bson and can be used to build a higher
+// level bson library.
+package bsoncore
+
+import (
+	"strconv"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+// DocumentBuilder builds a bson document
+type DocumentBuilder struct {
+	doc   []byte
+	index []int32
+}
+
+// ArrayBuilder builds a bson array
+type ArrayBuilder struct {
+	arr   []byte
+	index []int32
+	key   []int
```

Comment:
What's the purpose of the `key` slice?

Reply:
the key slice keeps the number of elements (the index) in the array. It's used to find the key when appending elements.